{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Initially, I relied on annotations from [Pfam](ftp://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.regions.uniprot.tsv.gz) and [Uniprot](ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/docs/pdbtosp.txt) (downloaded using the bash script [get_data.sh](./data/get_data.sh)), and focused on the [classic zinc finger domain](https://pfam.xfam.org/family/PF00096)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import os\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import re\n",
    "import subprocess as sp\n",
    "import tarfile\n",
    "from urllib import request\n",
    "\n",
    "# Initialize\n",
    "pfam_accession = \"PF00096\"\n",
    "prosite_id = \"PS50157\"\n",
    "CTCF = \"P49711\"\n",
    "\n",
    "#++++++++++++++++#\n",
    "# Pickle recipes #\n",
    "#++++++++++++++++#\n",
    "\n",
    "# Adapted from \"Load Faster in Python With Compressed Pickles\"\n",
    "# https://medium.com/better-programming/load-fast-load-big-with-compressed-pickles-5f311584507e\n",
    "\n",
    "def save_pickle(file_name, pkl):\n",
    "    \"\"\"\n",
    "    Saves a pickle.\n",
    "    \"\"\"\n",
    "\n",
    "    with gzip.open(file_name, \"wb\") as f:\n",
    "        pickle.dump(pkl, f)\n",
    "\n",
    "def load_pickle(file_name):\n",
    "    \"\"\"\n",
    "    Loads and returns a pickle.\n",
    "    \"\"\"\n",
    "\n",
    "    with gzip.open(file_name, \"rb\") as f:\n",
    "        pkl = pickle.load(f)\n",
    "\n",
    "    return(pkl)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I extracted the UniProt human proteome reference sequences (only one per gene) as [Seq](https://biopython.org/wiki/Seq) objects (*i.e.* `human_seqs`)..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20595\n"
     ]
    }
   ],
   "source": [
    "# Initialize\n",
    "human_seqs_pickle = \"./pkl/human_sequences.pkl.gz\"\n",
    "\n",
    "if not os.path.exists(human_seqs_pickle):\n",
    "\n",
    "    # Initialize\n",
    "    human_seqs = {}\n",
    "    pattern = re.compile(\"^\\w{2}\\|(\\w+)\")\n",
    "\n",
    "    with gzip.open(\"./data/UP000005640_9606.fasta.gz\", \"rt\") as f:\n",
    "        for seq_record in SeqIO.parse(f, \"fasta\"):\n",
    "            uniacc = pattern.match(seq_record.id)\n",
    "            human_seqs.setdefault(uniacc.group(1), seq_record.seq)\n",
    "\n",
    "    save_pickle(human_seqs_pickle, human_seqs)\n",
    "\n",
    "else:\n",
    "\n",
    "    human_seqs = load_pickle(human_seqs_pickle)\n",
    "\n",
    "# Sanity check:\n",
    "# According to UniProt's human reference proteome (https://www.uniprot.org/proteomes/UP000005640),\n",
    "# the gene count is 20,595\n",
    "print(len(human_seqs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "... and all zinc fingers, according to Pfam, mapped to human (and other UniProt) proteins (*i.e.* `uprot2pfam`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total number of C2H2-zf proteins in UniProt is: 160662\n",
      "For human, the number is: 678\n",
      "And for CTCF: 6\n"
     ]
    }
   ],
   "source": [
    "# Initialize\n",
    "uprot2pfam_pickle = \"./pkl/uniprot_to_pfam.pkl.gz\"\n",
    "\n",
    "if not os.path.exists(uprot2pfam_pickle):\n",
    "\n",
    "    # Initialize\n",
    "    uprot2pfam = {}\n",
    "\n",
    "    with gzip.open(\"./data/Pfam-A.regions.uniprot.tsv.gz\", \"rt\") as f:\n",
    "         for chunk in pd.read_csv(f, encoding=\"utf-8\", sep=\"\\t\", chunksize=1024):\n",
    "            for index, row in chunk.iterrows():\n",
    "                uniacc, seq_version, crc64, md5, pfam_acc, start, end = row.tolist()\n",
    "                if pfam_acc == pfam_accession:\n",
    "                    uprot2pfam.setdefault(uniacc, set())\n",
    "                    uprot2pfam[uniacc].add(tuple([start, end]))\n",
    "\n",
    "    save_pickle(uprot2pfam_pickle, uprot2pfam)\n",
    "\n",
    "else:\n",
    "\n",
    "    uprot2pfam = load_pickle(uprot2pfam_pickle)\n",
    "\n",
    "# Sanity checks:\n",
    "# The total number of C2H2-zf proteins in UniProt is?\n",
    "print(\"The total number of C2H2-zf proteins in UniProt is: %s\" % len(uprot2pfam))\n",
    "# According to \"The Human Transcription Factors\" (PMID: 29425488), the total C2H2-zf TFs is 747\n",
    "n = len(set(uprot2pfam.keys()).intersection(set(human_seqs.keys())))\n",
    "print(\"For human, the number is: %s\" % n)\n",
    "# According to \"Analysis of the vertebrate insulator protein CTCF-binding sites in the human genome\" (PMID: 17382889),\n",
    "# the total number of C2H2-zf domains for CTCF is 11\n",
    "print(\"And for CTCF: %s\" % len(uprot2pfam[CTCF]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Together, the fact that both i) the number of human C2H2-zf proteins and ii) CTCF C2H2-zf domains are lower than expected suggests that there might be something wrong with theÂ Pfam thresholding used by UniProt.\n",
    "\n",
    "Instead, I switched to PROSITE domain annotations by the [zinc finger C2H2 type domain profile](https://prosite.expasy.org/PS50157), which [match CTCF better](https://www.uniprot.org/uniprot/P49711).\n",
    "\n",
    "The file [uniprot2prosite.tab.gz](./data/uniprot2prosite.tab.gz) is obtained from [UniProt](https://www.uniprot.org/database/DB-0084) by customizing the output to `Entry`, `Gene names`, `Organism` and `PROSITE`.\n",
    "\n",
    "I extracted all C2H2-zf proteins from UniProt, according to PROSITE (*i.e.* `c2h2_zfs`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total number of C2H2-zf proteins in UniProt is: 353167\n",
      "And for human, the number is: 762\n"
     ]
    }
   ],
   "source": [
    "# Initialize\n",
    "c2h2_zf_pickle = \"./pkl/c2h2_zf.pkl.gz\"\n",
    "pattern = re.compile(prosite_id)\n",
    "\n",
    "if not os.path.exists(c2h2_zf_pickle):\n",
    "\n",
    "    # Initialize\n",
    "    c2h2_zfs = {}\n",
    "\n",
    "    with gzip.open(\"./data/uniprot2prosite.tab.gz\", \"rt\") as f:\n",
    "         for chunk in pd.read_csv(f, encoding=\"utf-8\", sep=\"\\t\", chunksize=1024):\n",
    "            for index, row in chunk.iterrows():\n",
    "                uniacc, gene_name, organism, prosites = row.tolist()\n",
    "                if pattern.search(prosites):\n",
    "                    c2h2_zfs.setdefault(organism, set())\n",
    "                    c2h2_zfs[organism].add(uniacc)\n",
    "\n",
    "    save_pickle(c2h2_zf_pickle, c2h2_zfs)\n",
    "\n",
    "else:\n",
    "\n",
    "    c2h2_zfs = load_pickle(c2h2_zf_pickle)    \n",
    "\n",
    "# Sanity checks:\n",
    "# The total number of C2H2-zf proteins in UniProt is?\n",
    "c2h2_zf_set = set()\n",
    "for f in c2h2_zfs:\n",
    "    c2h2_zf_set.update(c2h2_zfs[f])\n",
    "print(\"The total number of C2H2-zf proteins in UniProt is: %s\" % len(c2h2_zf_set))\n",
    "# According to \"The Human Transcription Factors\" (PMID: 29425488), the total C2H2-zf TFs is 747\n",
    "n = len(c2h2_zfs[\"Homo sapiens (Human)\"].intersection(set(human_seqs.keys())))\n",
    "print(\"And for human, the number is: %s\" % n)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, I retrieved the sequences (only one per gene) of all C2H2-zf proteins from the UniProt [reference proteomes of eukaryotes](ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/reference_proteomes/Reference_Proteomes_2020_02.tar.gz)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total number of C2H2-zf proteins in UniProt reference proteomes of eukaryotes is: 186703\n"
     ]
    }
   ],
   "source": [
    "# Initialize\n",
    "c2h2_zf_seqs_pickle = \"./pkl/c2h2_zf_sequences.pkl.gz\"\n",
    "\n",
    "if not os.path.exists(c2h2_zf_seqs_pickle):\n",
    "\n",
    "    # Initialize\n",
    "    c2h2_zf_seqs = {}\n",
    "    dummy_dir = \"./data/\"\n",
    "    pattern = re.compile(\"Eukaryota/UP\\d+_\\d+.fasta.gz\")\n",
    "    pattern2 = re.compile(\"^\\w{2}\\|(\\w+)\")\n",
    "    reference_proteomes_fasta = \"./data/Reference_Proteomes_2020_01.tar.gz\"\n",
    "\n",
    "    with tarfile.open(reference_proteomes_fasta, \"r:gz\") as tar:\n",
    "        for member in tar.getmembers():\n",
    "            # Skip organisms other than eukaryotes\n",
    "            if not member.name.startswith(\"Eukaryota\"):\n",
    "                continue\n",
    "            # Skip files other than reference proteomes\n",
    "            if pattern.search(member.name):\n",
    "                reference_proteome_file = os.path.join(dummy_dir, member.name) \n",
    "                if not os.path.exists(reference_proteome_file):\n",
    "                    tar.extract(member, path=dummy_dir)\n",
    "                with gzip.open(reference_proteome_file, \"rt\") as f:\n",
    "                    for seq_record in SeqIO.parse(f, \"fasta\"):\n",
    "                        uniacc = pattern2.match(seq_record.id)\n",
    "                        if uniacc.group(1) in c2h2_zf_set:\n",
    "                            c2h2_zf_seqs.setdefault(uniacc.group(1), str(seq_record.seq))\n",
    "\n",
    "    save_pickle(c2h2_zf_seqs_pickle, c2h2_zf_seqs)\n",
    "\n",
    "else:\n",
    "\n",
    "    c2h2_zf_seqs = load_pickle(c2h2_zf_seqs_pickle)\n",
    "\n",
    "# Total C2H2-zf transcription factors in reference proteomes of eukaryotes is?\n",
    "print(\"The total number of C2H2-zf proteins in UniProt reference proteomes of eukaryotes is: %s\" % len(c2h2_zf_seqs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, I scanned the retrieved sequences with the [zinc finger C2H2 type domain profile](https://prosite.expasy.org/PS50157) from PROSITE and, for each protein, I stored the positions of each match (*i.e.* `uprot2prosite`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total number of C2H2-zf domains in UniProt reference proteomes of eukaryotes is: 963553\n",
      "For human, the number is: 7180\n",
      "And for CTCF: 11\n"
     ]
    }
   ],
   "source": [
    "# Initialize\n",
    "uprot2prosite_pickle = \"./pkl/uniprot_to_prosite.pkl.gz\"\n",
    "\n",
    "def scan_prosite(uniacc, uniacc2seqs):\n",
    "\n",
    "    # Initialize\n",
    "    matches = set()\n",
    "    fasta_file = \"./data/%s.fasta\" % uniacc\n",
    "\n",
    "    if os.path.exists(fasta_file):\n",
    "        os.remove(fasta_file)\n",
    "\n",
    "    with open(fasta_file, \"w\") as o:\n",
    "        o.write(\">%s\\n%s\\n\" % (uniacc, uniacc2seqs[uniacc]))\n",
    "\n",
    "    # Run PROSITE\n",
    "    cmd = \"perl %s/ps_scan.pl -d %s/prosite.dat -p %s %s\" % (prosite_dir, prosite_dir, prosite_id, fasta_file)\n",
    "    p = sp.run([cmd], shell=True, stdout=sp.PIPE, stderr=sp.PIPE)\n",
    "    for line in p.stdout.decode(\"utf-8\").split(\"\\n\"):\n",
    "        if not line.startswith(\">\"):\n",
    "            match = re.findall(\"\\S+\", line)\n",
    "            if match:\n",
    "                matches.add(tuple([match[0], match[2]]))\n",
    "\n",
    "    os.remove(fasta_file)\n",
    "\n",
    "    return(uniacc, matches)\n",
    "\n",
    "if not os.path.exists(uprot2prosite_pickle):\n",
    "\n",
    "    # Initialize\n",
    "    uprot2prosite = {}\n",
    "    c2h2_zf_prosite_matches = \"./data/c2h2_zf_prosite_matches.txt\"\n",
    "\n",
    "    if not os.path.exists(c2h2_zf_prosite_matches):\n",
    "\n",
    "        # Initialize\n",
    "        prosite_dir = \"/space/home/oriol/Programs/ps_scan\"\n",
    "        c2h2_zf_seqs_fasta = \"./data/c2h2_zf_sequences.fasta\"\n",
    "\n",
    "        if not os.path.exists(c2h2_zf_seqs_fasta):\n",
    "            with open(c2h2_zf_seqs_fasta, \"w\") as o:\n",
    "                for uniacc in c2h2_zf_seqs:\n",
    "                    o.write(\">%s\\n%s\\n\" % (uniacc, c2h2_zf_seqs[uniacc]))\n",
    "\n",
    "        # Run PROSITE\n",
    "        cmd = \"perl %s/ps_scan.pl -d %s/prosite.dat -p %s %s > %s\" % (prosite_dir, prosite_dir, prosite_id,\n",
    "            c2h2_zf_seqs_fasta, c2h2_zf_prosite_matches)\n",
    "        os.system(cmd)\n",
    "\n",
    "    with open(c2h2_zf_prosite_matches) as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\">\"):\n",
    "                m = re.search(\">(\\S+) :\", line)\n",
    "                uniacc = m.group(1)\n",
    "                uprot2prosite.setdefault(uniacc, set())\n",
    "            else:\n",
    "                m = re.search(\"(\\d+) - (\\d+)\", line)\n",
    "                if m:\n",
    "                    uprot2prosite[uniacc].add(tuple([m.group(1), m.group(2)]))\n",
    "\n",
    "    save_pickle(uprot2prosite_pickle, uprot2prosite)\n",
    "\n",
    "else:\n",
    "\n",
    "    uprot2prosite = load_pickle(uprot2prosite_pickle)\n",
    "\n",
    "# Sanity checks:\n",
    "# The total number of C2H2-zf domains in UniProt reference proteomes of eukaryotes is?\n",
    "n = sum([len(uprot2prosite[uniacc]) for uniacc in uprot2prosite])\n",
    "print(\"The total number of C2H2-zf domains in UniProt reference proteomes of eukaryotes is: %s\" % n)\n",
    "# The total number of human C2H2-zf domains in UniProt reference proteomes of eukaryotes is?\n",
    "n = sum([len(uprot2prosite[uniacc]) for uniacc in set(uprot2prosite.keys()).intersection(set(human_seqs.keys()))])\n",
    "print(\"For human, the number is: %s\" % n)\n",
    "# According to \"Analysis of the vertebrate insulator protein CTCF-binding sites in the human genome\" (PMID: 17382889),\n",
    "# the total number of C2H2-zf domains for CTCF is 11\n",
    "print(\"And for CTCF: %s\" % len(uprot2prosite[CTCF]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The PROSITE results are more in line with my expectations. Hence, I will focus on the PROSITE-based C2H2-zf proteins and domains.\n",
    "\n",
    "The file [uniprot2pdb.tab.gz](./data/uniprot2pdb.tab.gz) is obtained from [UniProt](https://www.uniprot.org/database/DB-0070) by customizing the output to `Entry`, `Gene names`, `Organism` and `PDB`.\n",
    "\n",
    "I extracted all PDB accessions associated with C2H2-zf proteins from UniProt."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total number of C2H2-zf proteins with a 3D structure in the PDB is: 136\n",
      "For human, the number is: 91\n"
     ]
    }
   ],
   "source": [
    "# Initialize\n",
    "c2h2_zf_pdb_pickle = \"./pkl/c2h2_zf_pdb.pkl.gz\"\n",
    "\n",
    "if not os.path.exists(c2h2_zf_pdb_pickle):\n",
    "\n",
    "    # Initialize\n",
    "    c2h2_zfs_pdbs = {}\n",
    "\n",
    "    with gzip.open(\"./data/uniprot2pdb.tab.gz\", \"rt\") as f:\n",
    "         for chunk in pd.read_csv(f, encoding=\"utf-8\", sep=\"\\t\", chunksize=1024):\n",
    "            for index, row in chunk.iterrows():\n",
    "                uniacc, gene_name, organism, pdb_string = row.tolist()\n",
    "                if organism in c2h2_zfs:\n",
    "                    if uniacc in c2h2_zfs[organism]:\n",
    "                        pdb_list = pdb_string.split(\";\")\n",
    "                        p = pdb_list.pop()\n",
    "                        c2h2_zfs_pdbs.setdefault(uniacc, pdb_list)\n",
    "\n",
    "    save_pickle(c2h2_zf_pdb_pickle, c2h2_zfs_pdbs)\n",
    "\n",
    "else:\n",
    "\n",
    "    c2h2_zfs_pdbs = load_pickle(c2h2_zf_pdb_pickle) \n",
    "\n",
    "# The total number of C2H2-zf proteins with a 3D structure in the PDB is?\n",
    "print(\"The total number of C2H2-zf proteins with a 3D structure in the PDB is: %s\" % len(c2h2_zfs_pdbs))\n",
    "# The total number of human C2H2-zf proteins with a 3D structure in the PDB is?\n",
    "n = set(c2h2_zfs_pdbs.keys()).intersection(human_seqs.keys())\n",
    "print(\"For human, the number is: %s\" % len(n))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, I retrieved all protein-DNA accessions from the PDB."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total number of protein-DNA complexes in the PDB is: 5587\n",
      "For C2H2-zf proteins, the number is: 23\n",
      "And for human: 12\n",
      "* CTCF has 12 protein-DNA PDB complexes: 5K5H;5K5I;5K5J;5K5L;5KKQ;5T00;5T0U;5UND;5YEF;5YEG;5YEH;5YEL;\n"
     ]
    }
   ],
   "source": [
    "# Initialize\n",
    "pdb_protein_dna_pickle = \"./pkl/pdb_protein_dna.pkl.gz\"\n",
    "\n",
    "if not os.path.exists(pdb_protein_dna_pickle):\n",
    "\n",
    "    # Initialize\n",
    "    pattern = re.compile(\"\\w{4}\")\n",
    "    pdb_protein_dna = set()\n",
    "    cwd = os.getcwd()\n",
    "    url = \"http://www.rcsb.org/pdb/rest/search\"\n",
    "    xml = \"\".join(['<?xml version=\"1.0\" encoding=\"UTF-8\"?>',\n",
    "                   '<orgPdbQuery>',\n",
    "                   '<queryType>org.pdb.query.simple.ChainTypeQuery</queryType>',\n",
    "                   '<description>Chain Type Search : Contains Protein=Y, Contains DNA=Y, Contains RNA=I, Contains DNA/RNA Hybrid=I</description>',\n",
    "                   '<containsProtein>Y</containsProtein>',\n",
    "                   '<containsDna>Y</containsDna>',\n",
    "                   '<containsRna>I</containsRna>',\n",
    "                   '<containsHybrid>I</containsHybrid>',\n",
    "                   '</orgPdbQuery>'])\n",
    "\n",
    "    r = request.Request(url, data=xml.encode(\"utf-8\"))\n",
    "    for line in request.urlopen(r):\n",
    "        pdb = line.decode(\"utf-8\").strip()\n",
    "        if pattern.search(pdb):\n",
    "            pdb_protein_dna.add(pdb)\n",
    "\n",
    "    save_pickle(pdb_protein_dna_pickle, pdb_protein_dna)\n",
    "\n",
    "else:\n",
    "\n",
    "    pdb_protein_dna = load_pickle(pdb_protein_dna_pickle)\n",
    "\n",
    "# The total number of protein-DNA complexes in the PDB is?\n",
    "print(\"The total number of protein-DNA complexes in the PDB is: %s\" % len(pdb_protein_dna))\n",
    "# The total number of protein-DNA complexes of C2H2-zf proteins in the PDB is?\n",
    "n = 0\n",
    "for uniacc in c2h2_zfs_pdbs:\n",
    "    if set(c2h2_zfs_pdbs[uniacc]).intersection(pdb_protein_dna):\n",
    "        n += 1\n",
    "        if uniacc == CTCF:\n",
    "            CTCF_set = sorted(set(c2h2_zfs_pdbs[uniacc]).intersection(pdb_protein_dna))\n",
    "print(\"For C2H2-zf proteins, the number is: %s\" % n)\n",
    "# The total number of protein-DNA complexes of human C2H2-zf proteins in the PDB is?\n",
    "n = 0\n",
    "for uniacc in c2h2_zfs_pdbs:\n",
    "    if uniacc not in human_seqs:\n",
    "        continue\n",
    "    if set(c2h2_zfs_pdbs[uniacc]).intersection(pdb_protein_dna):\n",
    "        n += 1\n",
    "print(\"And for human: %s\" % n)\n",
    "print(\"* CTCF has %s protein-DNA PDB complexes: %s;\" % (len(CTCF_set), \";\".join(CTCF_set)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now here was the tricky part: How many (what percentage of) C2H2-zf domains are covered in the PDB?\n",
    "\n",
    "To answer this question, I used the [pdb2uniprot](https://github.com/mgalardini/pdb2uniprot) package developed by [Marco Galardini](https://github.com/mgalardini).\n",
    "\n",
    "First of all, I generated a list of PDB accessions associated with C2H2-zf proteins with which to run pdb2uniprot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize\n",
    "c2h2_zf_pdb_file = \"./data/c2h2_zf_pdb.txt\"\n",
    "\n",
    "if not os.path.exists(c2h2_zf_pdb_file):\n",
    "\n",
    "    with open(c2h2_zf_pdb_file, \"w\") as o:\n",
    "        for uniacc in c2h2_zfs_pdbs:\n",
    "            # i.e. lower is required by pdb2uniprot\n",
    "            o.write(\"%s\\n\" % \"\\n\".join([pdb.lower() for pdb in c2h2_zfs_pdbs[uniacc]]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The file [c2h2_zf_pdb2uniprot](\"./data/c2h2_zf_pdb2uniprot.tsv.gz\") stores the results from pdb2uniprot:\n",
    "\n",
    "```\n",
    "$ zless c2h2_zf_pdb2uniprot.tsv.gz | head\n",
    "pdb id\tpdb chain\tpdb residue\tpdb position\tuniprot id\tuniprot residue\tuniprot position\n",
    "6u04\tA\tA\t1\tP84243\tA\t2\n",
    "6u04\tA\tR\t2\tP84243\tR\t3\n",
    "6u04\tA\tT\t3\tP84243\tT\t4\n",
    "6u04\tA\tK\tnull\tP84243\tK\t5\n",
    "6u04\tA\tQ\tnull\tP84243\tQ\t6\n",
    "6u04\tA\tT\tnull\tP84243\tT\t7\n",
    "6u04\tA\tA\tnull\tP84243\tA\t8\n",
    "6u04\tA\tR\tnull\tP84243\tR\t9\n",
    "6u04\tA\tK\tnull\tP84243\tK\t10\n",
    "```\n",
    "\n",
    "Then, for each C2H2-zf PDB structure, I extracted the residues of the UniProt C2H2-zf protein covered. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12\n"
     ]
    }
   ],
   "source": [
    "# Initialize\n",
    "c2h2_zf_pdb2uniprot_pickle = \"./pkl/c2h2_zf_pdb2uniprot.pkl.gz\"\n",
    "\n",
    "if not os.path.exists(c2h2_zf_pdb2uniprot_pickle):\n",
    "\n",
    "    # Initialize\n",
    "    c2h2_zf_pdb2uniprot = {}\n",
    "\n",
    "    with gzip.open(\"./data/c2h2_zf_pdb2uniprot.tsv.gz\", \"rt\") as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"pdb id\"):\n",
    "                continue\n",
    "            pdb, pdb_chain, pdb_res, pdb_pos, uniacc, uni_res, uni_pos = line.strip().split(\"\\t\")\n",
    "            if pdb_res != \"null\":\n",
    "                pdb2uniprot = tuple([pdb, uniacc])\n",
    "                c2h2_zf_pdb2uniprot.setdefault(pdb2uniprot, {})\n",
    "                c2h2_zf_pdb2uniprot[pdb2uniprot].setdefault(pdb_chain, [])\n",
    "                c2h2_zf_pdb2uniprot[pdb2uniprot][pdb_chain].append(int(uni_pos))\n",
    "\n",
    "    save_pickle(c2h2_zf_pdb2uniprot_pickle, c2h2_zf_pdb2uniprot)\n",
    "\n",
    "else:\n",
    "\n",
    "    c2h2_zf_pdb2uniprot = load_pickle(c2h2_zf_pdb2uniprot_pickle)\n",
    "\n",
    "# Sanity checks:\n",
    "# The total number of protein-DNA complexes in the PDB for CTCF is 12\n",
    "n = set()\n",
    "for pdb, uniacc in c2h2_zf_pdb2uniprot:\n",
    "    if uniacc == CTCF and pdb.upper() in pdb_protein_dna:\n",
    "        n.add(pdb)\n",
    "print(len(n))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, I extracted the \"magic numbers\"!!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total number of C2H2-zf domains covered by a 3D structure in the PDB is: 331\n",
      "For human, the number is: 270\n",
      "Moreover, the total number of C2H2-zf domains covered by a protein-DNA complex in the PDB is: 86\n",
      "And for human, the number is: 48\n"
     ]
    }
   ],
   "source": [
    "def overlap(start_A, end_A, start_B, end_B):\n",
    "    \"\"\"\n",
    "    ---------AAAAAAA---------\n",
    "    -----BBBBB--------------- = True\n",
    "    ----------BBBBB---------- = True\n",
    "    ---------------BBBBB----- = True\n",
    "    ----BBBBB---------------- = False\n",
    "    ----------------BBBBB---- = False\n",
    "    \"\"\"\n",
    "\n",
    "    if start_A <= end_B and end_A >= start_B:\n",
    "\n",
    "        return(True)\n",
    "\n",
    "# The total number of C2H2-zf domains covered by a 3D structure in the PDB is?\n",
    "n = set()\n",
    "m = set()\n",
    "pdbs = {}\n",
    "for pdb, uniacc in c2h2_zf_pdb2uniprot:\n",
    "    if uniacc in uprot2prosite:\n",
    "        for domain in uprot2prosite[uniacc]:\n",
    "            domain_start = int(domain[0])\n",
    "            domain_end = int(domain[1])\n",
    "            for pdb_chain in c2h2_zf_pdb2uniprot[(pdb, uniacc)]:\n",
    "                pdb_chain_start = c2h2_zf_pdb2uniprot[(pdb, uniacc)][pdb_chain][0]\n",
    "                pdb_chain_end = c2h2_zf_pdb2uniprot[(pdb, uniacc)][pdb_chain][-1]\n",
    "                if overlap(domain_start, domain_end, pdb_chain_start, pdb_chain_end):\n",
    "                    n.add(tuple([uniacc, domain]))\n",
    "                    if pdb.upper() in pdb_protein_dna:\n",
    "                        m.add(tuple([uniacc, domain]))\n",
    "                        pdbs.setdefault(pdb.upper(), uniacc)\n",
    "print(\"The total number of C2H2-zf domains covered by a 3D structure in the PDB is: %s\" % len(n))\n",
    "# The total number of human C2H2-zf domains covered by a 3D structure in the PDB is?\n",
    "i = set()\n",
    "for u, d in n:\n",
    "    if u in human_seqs:\n",
    "        i.add(tuple([u, d]))\n",
    "print(\"For human, the number is: %s\" % len(i))\n",
    "# The total number of C2H2-zf domains covered by a protein-DNA complex in the PDB is?\n",
    "print(\"Moreover, the total number of C2H2-zf domains covered by a protein-DNA complex in the PDB is: %s\" % len(m))\n",
    "# The total number of human C2H2-zf domains covered by a protein-DNA complex in the PDB is?\n",
    "i = set()\n",
    "for u, d in m:\n",
    "    if u in human_seqs:\n",
    "        i.add(tuple([u, d]))\n",
    "print(\"And for human, the number is: %s\" % len(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total number of protein-DNA complexes in the PDB covering one or more C2H2-zf domains is: 104\n",
      "For human the number is: 62\n",
      "The total number of C2H2-zf proteins with one or more domains covered by protein-DNA complexes in the PDB is: 19\n",
      "For human, the number is: 11\n",
      "\n",
      "Must have list of C2H2-zf protein-DNA complexes:\n",
      "1A1F\n",
      "1A1G\n",
      "1A1H\n",
      "1A1I\n",
      "1A1J\n",
      "1A1K\n",
      "1A1L\n",
      "1AAY\n",
      "1F2I\n",
      "1G2D\n",
      "1G2F\n",
      "1JK1\n",
      "1JK2\n",
      "1LLM\n",
      "1P47\n",
      "1UBD\n",
      "1ZAA\n",
      "2GLI\n",
      "2I13\n",
      "2JP9\n",
      "2JPA\n",
      "2KMK\n",
      "2LT7\n",
      "2PRT\n",
      "2WBS\n",
      "2WBU\n",
      "3UK3\n",
      "4F2J\n",
      "4F6M\n",
      "4F6N\n",
      "4GZN\n",
      "4IS1\n",
      "4M9E\n",
      "4M9V\n",
      "4R2A\n",
      "4R2C\n",
      "4R2D\n",
      "4R2E\n",
      "4R2P\n",
      "4R2Q\n",
      "4R2R\n",
      "4R2S\n",
      "4X9J\n",
      "5EGB\n",
      "5EH2\n",
      "5EI9\n",
      "5K5H\n",
      "5K5I\n",
      "5K5J\n",
      "5K5L\n",
      "5KE6\n",
      "5KE7\n",
      "5KE8\n",
      "5KE9\n",
      "5KEA\n",
      "5KEB\n",
      "5KKQ\n",
      "5KL2\n",
      "5KL3\n",
      "5KL4\n",
      "5KL5\n",
      "5KL6\n",
      "5KL7\n",
      "5T00\n",
      "5T0U\n",
      "5UND\n",
      "5V3J\n",
      "5V3M\n",
      "5VMU\n",
      "5VMV\n",
      "5VMW\n",
      "5VMX\n",
      "5VMY\n",
      "5VMZ\n",
      "5WJQ\n",
      "5YEF\n",
      "5YEG\n",
      "5YEH\n",
      "5YEL\n",
      "5YJ3\n",
      "6A57\n",
      "6B0O\n",
      "6B0P\n",
      "6B0Q\n",
      "6B0R\n",
      "6BLW\n",
      "6DF5\n",
      "6DF8\n",
      "6DF9\n",
      "6DFA\n",
      "6DFB\n",
      "6DFC\n",
      "6E93\n",
      "6E94\n",
      "6JNL\n",
      "6JNM\n",
      "6JNN\n",
      "6KI6\n",
      "6ML2\n",
      "6ML3\n",
      "6ML4\n",
      "6ML5\n",
      "6ML6\n",
      "6ML7\n"
     ]
    }
   ],
   "source": [
    "# Must have list of PDB accessions (i.e. protein-DNA complexes of C2H2-zf domains)\n",
    "print(\"The total number of protein-DNA complexes in the PDB covering one or more C2H2-zf domains is: %s\" % len(pdbs))\n",
    "print(\"For human the number is: %s\" % len([p for p in pdbs if pdbs[p] in set(pdbs.values()).intersection(set(human_seqs.keys()))]))\n",
    "print(\"The total number of C2H2-zf proteins with one or more domains covered by protein-DNA complexes in the PDB is: %s\" % len(set(pdbs.values())))\n",
    "print(\"For human, the number is: %s\" % len(set(pdbs.values()).intersection(set(human_seqs.keys()))))\n",
    "print()\n",
    "print(\"Must have list of C2H2-zf protein-DNA complexes:\")\n",
    "for pdb in sorted(pdbs):\n",
    "    print(pdb)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
